src/stream/KeyExchange.js

```diff
@@ -96,29 +97,47 @@ function GroupKeyStore({ groupKeys = new Map() }) {
         has(groupKeyId) {
             if (currentGroupKeyId === groupKeyId) { return true }
 
-            if (nextGroupKey && nextGroupKey.id === groupKeyId) { return true }
+            if (nextGroupKeys.some((nextKey) => nextKey.id === groupKeyId)) { return true }
 
             return store.has(groupKeyId)
         },
         isEmpty() {
-            return !nextGroupKey && store.size === 0
+            return nextGroupKeys.length === 0 && store.size === 0
         },
         useGroupKey() {
-            if (nextGroupKey) {
-                // next key becomes current key
+            const nextGroupKey = nextGroupKeys.pop()
+            // first message
+            if (!currentGroupKeyId && nextGroupKey) {
```

Comment:
To me this would be more readable if there was only one check in each `if` statement and some else statements. It would help to see the overall structure of the method (that we do exactly one of four possible things). 
```
if (nextGroupKey) {
    storeKey(nextGroupKey)
    if (currentGroupKeyId) {
        // key changed
    } else {
        // first message
    }
} else {
    if (currentGroupKeyId) {
        // ...
    } else {
        // generate & use key if none already set
    }
}
```

Reply:
I have a strong distaste for following logic through `if`/`else` statements, and even more so nested `if`/`else`.

What do you think about a `switch` statement instead?

```js
switch (true) {
    // first message
    case (!currentGroupKeyId && nextGroupKey): {
        return […]
    }
    // key changed
    case (currentGroupKeyId && nextGroupKey): {
        return […]
    }
    // generate & use key if none already set
    case (!currentGroupKeyId): {
        return […]
    }
    default: {
        return […]
    }
}
```

Reply:
Nice, that communicates the structure cleverly! 👍